docs/decisions/0045-updated-vector-store-design.md

```diff
@@ -0,0 +1,811 @@
+{
+}
+
+```
+
+#### Decision Outcome
+
+Chosen option: 4 + 5.
+
+- Collection create and configuration varies considerably across different schemas, even for the same database, but also varies by database type.
+- Collection list, exists and delete is the same across different schemas, but varies by database type.
+- Vector storage, even with custom schemas can be supported using a single implementation per database type.
+- We will therefore need to support multiple collection create service implementations per store type, a single collection update service implementation per store type, and a single vector store implementation per store type.
+- At the same time we can layer interfaces on top that allow easy combined access to collection and record management.
+
+
+###  Question 2: Collection name and key value normalization in store, decorator or via injection.
+
+#### Option 1 - Normalization in main record service
+
+- Pros: Simple
+- Cons: The normalization needs to vary separately from the record service, so this will not work
+
+```cs
+    public class AzureAISearchMemoryRecordService<TDataModel> : IMemoryRecordService<TDataModel>
+    {
+        ...
+
+        // On input.
+        var normalizedCollectionName = this.NormalizeCollectionName(collectionName);
+        var encodedId = AzureAISearchMemoryRecord.EncodeId(key);
+
+        ...
+
+        // On output.
+        DecodeId(this.Id)
+
+        ...
+    }
+```
+
+#### Option 2 - Normalization in decorator
+
+- Pros: Allows normalization to vary separately from the record service.
+- Pros: No code executed when no normalization required.
+- Pros: Easy to package matching encoders/decoders together.
+- Pros: Easier to obsolete encoding/normalization as a concept.
+- Cons: Not a major con, but need to implement the full MemoryRecordService interface, instead of e.g. just providing the two translation functions, if we go with option 3.
+- Cons: Hard to have a generic implementation that can work with any model, without either changing the data in the provided object on upsert or doing cloning in an expensive way.
+
+```cs
+    new KeyNormalizingAISearchMemoryRecordService<MyModel>(
+        "keyField",
+         new AzureAISearchMemoryRecordService<MyModel>(...));
+```
+
+#### Option 3 - Normalization via optional function parameters to record service constructor
+
+- Pros: Allows normalization to vary separately from the record service.
+- Pros: No need to implement the full MemoryRecordService interface.
+- Pros: Can modify values on serialization without changing the incoming record, if supported by DB SDK.
+- Cons: Harder to package matching encoders/decoders together.
+
+```cs
+public class AzureAISearchMemoryRecordService<TDataModel>(StoreOptions options);
+
+public class StoreOptions
+{
+    public Func<string, string>? EncodeKey { get; init; }
+    public Func<string, string>? DecodeKey { get; init; }
+    public Func<string, string>? SanitizeCollectionName { get; init; }
```

Comment:
I'm wondering if we need normalization logic only for key and collection name or for other fields as well? Maybe we should introduce some optional field mapper/transformer that will work for any property?

The great thing about custom schema and vector DB abstractions is the ability to use any of vector DB implementations with the same schema and easily switch between them. Just like `Key` property format may be compatible with Qdrant but not with Azure AI Search, similar situations may happen when supporting different property types, especially datetimes, boolean flags (previously there was an issue with some connector that used `0` and `1` for boolean and then switched to `true`/`false`). It feels like optional mapper should handle any data model property. This should provide more flexibility for users.